Gems/PhysX/Code/Editor/Source/ComponentModes/Joints/JointsComponentMode.cpp

```diff
@@ -0,0 +1,598 @@
+        if (m_subModes[m_subMode])
+        {
+            m_subModes[m_subMode]->HandleMouseInteraction(mouseInteraction);
+        }
+
+        return false;
+    }
+
+    void JointsComponentMode::SetupSubModes(const AZ::EntityComponentIdPair& entityComponentIdPair)
+    {
+        //create the 3 cluster groups
+        for (auto& clusterId : m_modeSelectionClusterIds)
+        {
+            AzToolsFramework::ViewportUi::ViewportUiRequestBus::EventResult(
+                clusterId, AzToolsFramework::ViewportUi::DefaultViewportId,
+                &AzToolsFramework::ViewportUi::ViewportUiRequestBus::Events::CreateCluster,
+                AzToolsFramework::ViewportUi::Alignment::TopLeft);
+        }
+
+        //retrieve the enabled sub components from the entity
+        AZStd::vector<JointsComponentModeCommon::SubModeParamaterState> subModesState;
+        EditorJointRequestBus::EventResult(subModesState, entityComponentIdPair, &EditorJointRequests::GetSubComponentModesState);
+
+        const AzToolsFramework::ViewportUi::ClusterId group1ClusterId = GetClusterId(ClusterGroups::Group1);
+        const AzToolsFramework::ViewportUi::ClusterId group2ClusterId = GetClusterId(ClusterGroups::Group2);
+        //hide cluster 2, if something is added to it. it will make is visible
+        AzToolsFramework::ViewportUi::ViewportUiRequestBus::Event(
+            AzToolsFramework::ViewportUi::DefaultViewportId, &AzToolsFramework::ViewportUi::ViewportUiRequestBus::Events::SetClusterVisible,
+            group2ClusterId, false);
+
+        const AzToolsFramework::ViewportUi::ClusterId group3ClusterId = GetClusterId(ClusterGroups::Group3);
+        // hide cluster 3, if something is added to it. it will make is visible
+        AzToolsFramework::ViewportUi::ViewportUiRequestBus::Event(
+            AzToolsFramework::ViewportUi::DefaultViewportId, &AzToolsFramework::ViewportUi::ViewportUiRequestBus::Events::SetClusterVisible,
+            group3ClusterId, false);
+
+        //translation and rotation are enabled for all joints in group 1
+        m_subModes[JointsComponentModeCommon::SubComponentModes::ModeType::Translation] =
+            AZStd::make_unique<JointsSubComponentModeTranslation>();
+        m_buttonData[JointsComponentModeCommon::SubComponentModes::ModeType::Translation] =
+            ButtonData{ group1ClusterId, Internal::RegisterClusterButton(group1ClusterId, "Move", SubModeData::TranslationToolTip) };
+        
+        m_subModes[JointsComponentModeCommon::SubComponentModes::ModeType::Rotation] = AZStd::make_unique<JointsSubComponentModeRotation>();
+        m_buttonData[JointsComponentModeCommon::SubComponentModes::ModeType::Rotation] =
+            ButtonData{ group1ClusterId, Internal::RegisterClusterButton(group1ClusterId, "Rotate", SubModeData::RotationTitle) };
+
+        //some constants for some modes
+        const float exponentBreakage = 1.0f;
+        const float exponentSpring = 2.0f;
```

Comment:
nit: constants should ideally be `UpperCamel` case to separate them from variables (even at function scope)